lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +682,1178 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+      } else if (CalleeType == FunctionTypeValue::NoReturn) {
+#ifdef NORETURN_IS_FIXED
+        auto *Abort = CallInst::Create(M.getFunction("abort"));
+        new UnreachableInst(Term->getContext(), BB);
+        ReplaceInstWithInst(Term, Abort);
+#else
+        auto *Br = BranchInst::Create(getFallthrough(Term));
+        ReplaceInstWithInst(Term, Br);
+#endif
+      }
+
+      // Fix function_call marker
+      unsigned ArgNo = 0;
+      auto *I8PtrTy = Type::getInt8PtrTy(M.getContext());
+      auto *I8NullPtr = ConstantPointerNull::get(I8PtrTy);
+      for (Value *Arg : CI->args()) {
+        if (isa<BlockAddress>(Arg)) {
+          CI->setArgOperand(ArgNo, I8NullPtr);
+          if (Arg->use_empty())
+            cast<BlockAddress>(Arg)->destroyConstant();
+        }
+        ++ArgNo;
+      }
+
+      CallMap.insert({ CI, PCCallee });
+    }
+  }
+
+  // Extract outlined function
+  CodeExtractorAnalysisCache CEAC(*Root);
+  Function *OutlinedFunction = CodeExtractor(BlocksToExtract)
+                                 .extractCodeRegion(CEAC);
+
+  revng_assert(OutlinedFunction != nullptr);
+  revng_assert(OutlinedFunction->arg_size() == 0);
+  revng_assert(OutlinedFunction->getReturnType()->isVoidTy());
+
+  auto It = OutlinedFunction->user_begin();
+  cast<Instruction>(*It)->getParent()->eraseFromParent();
+
+  // Integrate function callee
+  for (auto &BB : *OutlinedFunction) {
+    if (isFunctionCall(&BB)) {
+      auto *Term = BB.getTerminator();
+      auto CalleePC = CallMap[getFunctionCall(Term)];
+      integrateFunctionCallee(&BB, CalleePC);
+
+      // Fix `unexpectedpc` callee when function is found to be Fake
+      if (Oracle.isFakeFunction(CalleePC)) {
+        for (auto &I : instructions(OutlinedFunction)) {
+          if (auto *Call = getCallTo(&I, UnexpectedPCMarker)) {
+            // TODO: fix if `unexpectedpc` does not exist in the caller (is that possible?)
+            revng_assert(UnexpectedPCCloned != nullptr);
+
+            auto *Br = BranchInst::Create(UnexpectedPCCloned);
+            ReplaceInstWithInst(I.getParent()->getTerminator(), Br);
+            Call->eraseFromParent();
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  // Make sure `newpc` is still FirstNonPhi, that would break further
+  // assumptions.
+  BasicBlock *Split = nullptr;
+  for (auto &BB : *OutlinedFunction) {
+    if (Split == &BB)
+      continue;
+
+    Split = nullptr;
+    for (auto &I : BB) {
+      if (auto *Call = getCallTo(&I, "newpc")) {
+        if (BB.getFirstNonPHI() != Call
+            && getLimitedValue(Call->getArgOperand(2)) == 1) {
+          if (isCallTo(Call->getPrevNode(), "function_call"))
+            Split = BB.splitBasicBlock(Call->getPrevNode(),
+                                       BB.getName() + Twine("__jt"));
+          else
+            Split = BB.splitBasicBlock(Call, BB.getName() + Twine("__jt"));
+          break;
+        }
+      }
+    }
+  }
```

Comment:
The way the `Split` is used is quite suspicious.
Can you describe what are you trying to achieve?

Reply:
Elaborated.